sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,278 @@
+        me); // callback context - i.e. user data.
+  }
+
+  if (me->mosquitto_handle == NULL)
+  {
+    ret = AZ_ERROR_OUT_OF_MEMORY;
+    return ret;
+  }
+
+  /* Configure callbacks. This should be done before connecting ideally. */
+  mosquitto_log_callback_set(me->mosquitto_handle, _az_mosquitto_on_log);
+  mosquitto_connect_callback_set(me->mosquitto_handle, _az_mosquitto_on_connect);
+  mosquitto_disconnect_callback_set(me->mosquitto_handle, _az_mosquitto_on_disconnect);
+  mosquitto_publish_callback_set(me->mosquitto_handle, _az_mosquitto_on_publish);
+  mosquitto_subscribe_callback_set(me->mosquitto_handle, _az_mosquitto_on_subscribe);
+  mosquitto_unsubscribe_callback_set(me->mosquitto_handle, _az_mosquitto_on_unsubscribe);
+  mosquitto_message_callback_set(me->mosquitto_handle, _az_mosquitto_on_message);
+
+  if (az_span_size(me->_internal.options.certificate_authority_trusted_roots) != 0)
+  {
+    _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_tls_set(
+        me->mosquitto_handle,
+        (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots),
+        NULL,
+        (const char*)az_span_ptr(connect_data->certificate.cert),
+        (const char*)az_span_ptr(connect_data->certificate.key),
+        NULL)));
+  }
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_username_pw_set(
+      me->mosquitto_handle,
+      (const char*)az_span_ptr(connect_data->username),
+      (const char*)az_span_ptr(connect_data->password))));
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_connect_async(
+      (struct mosquitto*)me->mosquitto_handle,
+      (char*)az_span_ptr(connect_data->host),
+      connect_data->port,
+      AZ_MQTT_KEEPALIVE_SECONDS)));
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_loop_start(me->mosquitto_handle)));
+
+  return ret;
+}
+
+AZ_NODISCARD az_result
+az_mqtt_outbound_sub(az_mqtt* mqtt, az_context* context, az_mqtt_sub_data* sub_data)
+{
+  (void)context;
+  return _az_result_from_mosq(mosquitto_subscribe(
+      mqtt->mosquitto_handle,
+      &sub_data->out_id,
+      (char*)az_span_ptr(sub_data->topic_filter),
+      sub_data->qos));
+}
+
+AZ_NODISCARD az_result
+az_mqtt_outbound_pub(az_mqtt* mqtt, az_context* context, az_mqtt_pub_data* pub_data)
+{
+  (void)context;
+  return _az_result_from_mosq(mosquitto_publish(
+      mqtt->mosquitto_handle,
+      &pub_data->out_id,
+      (char*)az_span_ptr(pub_data->topic), // Assumes properly formed NULL terminated string.
+      az_span_size(pub_data->payload),
+      az_span_ptr(pub_data->payload),
+      pub_data->qos,
+      false));
+}
+
+AZ_NODISCARD az_result az_mqtt_outbound_disconnect(az_mqtt* mqtt, az_context* context)
+{
+  (void)context;
+  return _az_result_from_mosq(mosquitto_disconnect(mqtt->mosquitto_handle));
+}
+
+AZ_NODISCARD az_result az_mqtt_wait_for_event(az_mqtt* mqtt, int32_t timeout)
+{
+  return _az_result_from_mosq(mosquitto_loop(mqtt->mosquitto_handle, timeout, 1));
```

Comment:
It looks like you shouldn't be using mosquitto_loop with mosquitto_connect_async
https://mosquitto.org/api/files/mosquitto-h.html#mosquitto_connect_async

Reply:
Also I don't think you should need to use mosquitto_loop if you're using mosquitto_loop_start

Reply:
Good point! This is to support the blocking (SYNC) mode which we decided to remove for now (originally under a separate compilation flag).

@RLeclair please remove this function - it shouldn't be used. We may decide to add it back if there is any demand for a blocking / "DoEvents" type of API interface.

Reply:
Removed!